### PR TITLE
New spirv

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

**Comment by kinke** [comment]
The path of the root dir inside the archive is `./_CPack_Packages/Darwin/TBZ2/LLVM-6.0.0git-18708be-Darwin`, but should be `LLVM-6.0.0`.

**Comment by thewilsonator** [comment]
Ah, I ran `tar -c ./_CPack_Packages/Darwin/TBZ2/LLVM-6.0.0git-18708be-Darwin | bzip2 -9 > ./_CPack_Packages/Darwin/TBZ2/LLVM-6.0.0git-18708be-Darwin.tar.bz2` from the llvm build directory instead of `cd`ing into `./_CPack_Packages/Darwin/TBZ2` and then running `tar`. It is very annoying that `libFuzzer.a`is not copied properly by cpack.

Will reupload tomorrow morning.

**Comment by thewilsonator** [comment]
Hmm it does not find `llvm-config`? It definitely there.

**Comment by kinke** [comment]
Most likely because of [this](https://github.com/ldc-developers/ldc/blob/master/cmake/Modules/FindLLVM.cmake#L30) missing 6.0 (and other future versions). Stumbled upon this yesterday.

**Comment by thewilsonator** [comment]
Hmm, still does not work.

**Comment by kinke** [comment]
You've used `./llvm-...`. Try repackaging without `.` (or simply try the extraction command locally before uploading ;)).

File: driver/cl_options.cpp [hunk 1]
```diff
@@ -343,10 +343,15 @@ cl::opt<llvm::Reloc::Model> mRelocModel(
...
     "code-model", cl::desc("Code model"), cl::ZeroOrMore,
+#if LDC_LLVM_VER < 600
     cl::init(llvm::CodeModel::Default),
     clEnumValues(
         clEnumValN(llvm::CodeModel::Default, "default",
                    "Target default code model"),
+#else
+    cl::init(llvm::CodeModel::Small),
+    clEnumValues(
+#endif
```
Comment (JohanEngelen) [inline driver/cl_options.cpp]: We shouldn't enforce CodeModel::Small.
The better solution is what LLVM does:
```
+cl::opt<llvm::CodeModel::Model> CMModel(
+    "code-model", cl::desc("Choose code model"),
+    cl::values(clEnumValN(CodeModel::Small, "small", "Small code model"),
+               clEnumValN(CodeModel::Kernel, "kernel", "Kernel code model"),
+               clEnumValN(CodeModel::Medium, "medium", "Medium code model"),
+               clEnumValN(CodeModel::Large, "large", "Large code model")));
+
+static inline Optional<CodeModel::Model> getCodeModel() {
+  if (CMModel.getNumOccurrences()) {
+    CodeModel::Model M = CMModel;
+    return M;
+  }
+  return None;
+}
```
mCodeModel should be made private and the getCodeModel function should be added.
Same as `getRelocModel` (misplaced in main.cpp)

Comment (thewilsonator) [inline driver/cl_options.cpp]: Will do. Ill just make sure this works.

Comment (kinke) [inline driver/cl_options.cpp]: This will all be superseded by #2148 anyway.

Comment (thewilsonator) [inline driver/cl_options.cpp]: Alright I'll leave it for #2148.

**Comment by thewilsonator** [comment]
Hmm, it timed out downloading LLVM. Closing/reopening.

**Comment by thewilsonator** [comment]
Dammit, I forgot to copy across `libFuzzer.a`. Is there a way to get it to do that automatically with cpack?

**Comment by thewilsonator** [comment]
A few sanitiser fuzz tests fail. Some due to missing line info:

```
/Users/nicholaswilson/d/ldc/tests/sanitizers/asan_interface.d:26:17: error: expected string not found in input
 // CHECK-NEXT: #0 {{.*}} in {{.*getX.*}} {{.*}}asan_interface.d:[[@LINE+1]]
                ^
<stdin>:3:16: note: scanning from here
READ of size 4 at 0x00010b523704 thread T0
               ^
<stdin>:3:16: note: with expression "@LINE+1" equal to "27"
READ of size 4 at 0x00010b523704 thread T0
               ^
<stdin>:4:78: note: possible intended match here
 #0 0x1078bf83f in _D14asan_interface4getXFZi (/Users/nicholaswilson/d/ldcbuild/tests/sanitizers/Output/asan_interface.d.tmp:x86_64+0x10000183f)
```
Others fail to link
```
$ "/Users/nicholaswilson/d/ldcbuild/bin/ldc2" "-g" "-fsanitize=fuzzer" "/Users/nicholaswilson/d/ldc/tests/sanitizers/fuzz_basic.d" "-of=/Users/nicholaswilson/d/ldcbuild/tests/sanitizers/Output/fuzz_basic.d.tmp"
# command stderr:
Undefined symbols for architecture x86_64:
  "___sanitizer_cov_trace_const_cmp1", referenced from:
      _LLVMFuzzerTestOneInput in fuzz_basic.d.o
  "___sanitizer_cov_trace_const_cmp4", referenced from:
      __D10fuzz_basic6FuzzMeFxPhmZv in fuzz_basic.d.o
  "___sanitizer_cov_trace_const_cmp8", referenced from:
      __D10fuzz_basic6FuzzMeFxPhmZv in fuzz_basic.d.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: /usr/bin/gcc failed with status: 1
```
The other failure is 
```
/Users/nicholaswilson/d/ldc/tests/sanitizers/fsanitize_fuzzer.d:13:12: error: expected string not found in input
 // CHECK: call {{.*}}_sanitizer_cov_trace_cmp
           ^
<stdin>:32:43: note: scanning from here
 call void @__sanitizer_cov_trace_pc_guard(i32* getelementptr inbounds ([6 x i32], [6 x i32]* @__sancov_gen_, i32 0, i32 0))
                                          ^
<stdin>:34:2: note: possible intended match here
 call void @__sanitizer_cov_trace_const_cmp8(i64 2, i64 %dataSize_arg)
 ^

error: command failed with exit status: 1
```

**Comment by thewilsonator** [comment]
Merging this so that testing and development can continue

**Comment by JohanEngelen** [comment]
@thewilsonator The link failure is likely due to versions not being up to date (LLVM vs. libFuzzer.a). Did you svn update projects/compiler-rt too?   In any case, just remove libFuzzer.a from the package for now. It's already more than enough pain to be testing LLVM trunk on Travis because it is failing so often.

**Comment by thewilsonator** [comment]
OK.

**Comment by JohanEngelen** [comment]
(libFuzzer was moved to compiler-rt, so for LLVM >= 6 we need to look for it in the same place as the sanitizer libs. Perhaps the version problem happened because you had an old libFuzzer lying around in the old location from a build before the move happened)

**Comment by kinke** [comment]
A new libfuzzer test also fails for LLVM 5.0 on Circle: `sanitizers/fuzz_basic.d`, see https://circleci.com/gh/ldc-developers/ldc/3051.

**Comment by JohanEngelen** [comment]
That's a sporadic error, let me know if it happens again, I will simplify the test.

**Pull request merged**